### PR TITLE
8335977: Deoptimization fails with assert "object should be reallocated already"

### PR DESCRIPTION
**Pull request by JohnTortugo**
Please, review this patch to fix an issue that may occur when serializing debug information related to reduce allocation merges. The problem happens when there are more than one JVMS in a `uncommon_trap` and a _younger_ JVMS doesn't have the RAM inputs as a local/expression/monitor but an older JVMS does. In that situation the loop at line 1173 of output.cpp will set the `is_root` property of the ObjectValue to `false` when processing the younger JVMS even though it may have been set to `true` when visiting the older JVMS.

Tested on: 
 - Win, Mac & Linux tier1-4 on x64 & Aarch64.
 - CTW with some thousands of jars.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335977](https://bugs.openjdk.org/browse/JDK-8335977): Deoptimization fails with assert "object should be reallocated already" (**Bug** - P2)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) Review applies to [8e345b59](https://git.openjdk.org/jdk/pull/21624/files/8e345b593a3553cded11103c8912164fc2d7a090)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Contributors
 * Christian Hagedorn `<chagedorn@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21624/head:pull/21624` \
`$ git checkout pull/21624`

Update a local copy of the PR: \
`$ git checkout pull/21624` \
`$ git pull https://git.openjdk.org/jdk.git pull/21624/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21624`

View PR using the GUI difftool: \
`$ git pr show -t 21624`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21624.diff">https://git.openjdk.org/jdk/pull/21624.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21624#issuecomment-2427669633)
</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/escapeAnalysis/TestReduceAllocationAndJVMStates.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        bFld = false;
+
+        for (int i = 0; i < 10000; i++) {
+            test(i % 2 == 0);
+        }
+        bFld = true;
+
+        // This will trigger a deoptimization which
+        // will make the issue manifest to the user
+        //test1(true);
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/escapeAnalysis/TestReduceAllocationAndJVMStates.java:53]: Should be removed?

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
The fix looks good to me.

Now that the `-XX:+StressUnstableIfTraps` option was added with [JDK-8335334](https://bugs.openjdk.org/browse/JDK-8335334), could you please also add [Test.java](https://bugs.openjdk.org/secure/attachment/111040/Test.java) that I used originally to reproduce the issue?

Does your test also reproduce the NPE example that @chhagedorn triggered with his [Test3.java](https://bugs.openjdk.org/secure/attachment/111067/Test3.java)? If not, I think that one should be added as well.

The more tests we have for this, the better :)

**Comment by TobiHartmann** [comment]
I executed some extended testing. All green.

**Comment by JohnTortugo** [comment]
> Does your test also reproduce the NPE example that @chhagedorn triggered with his [Test3.java](https://bugs.openjdk.org/secure/attachment/111067/Test3.java)? If not, I think that one should be added as well.

The test I added in this PR is based on @chhagedorn Test3.java. I was able to reproduce the issue on my end fairly easily.

> Now that the -XX:+StressUnstableIfTraps option was added with [JDK-8335334](https://bugs.openjdk.org/browse/JDK-8335334), could you please also add [Test.java](https://bugs.openjdk.org/secure/attachment/111040/Test.java) that I used originally to reproduce the issue?

I wasn't able to reproduce the issue using Test.java, how often does it reproduce for you with the flags that you listed at the top of Test.java ?

Thanks for testing @TobiHartmann !

File: src/hotspot/share/opto/output.cpp [hunk 2]
```diff
@@ -1158,6 +1172,12 @@ void PhaseOutput::Process_OopMap_Node(MachNode *mach, int current_offset) {
 
         for (int j = 0; j< merge->possible_objects()->length(); j++) {
           ObjectValue* ov = merge->possible_objects()->at(j)->as_ObjectValue();
+          // Already flagged as 'root' by something else. We shouldn't change it
+          // to non-root in a younger JVMS because it may need to be alive in
+          // the younger JVMS.
+          if (ov->is_root()) {
+            continue;
```
Comment (iwanowww) [inline src/hotspot/share/opto/output.cpp:1179]: You can either fuse `ov->is_root()` check into `is_root` computation (`bool is_root = ov->is_root() || ...`) or turn it into an `if-then-else` (`if (ov->is_root()) { /* comment */ } else { bool is_root = ...;  ov->set_root(is_root); }`). I find both cases easier to read.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by TobiHartmann** [comment]
> The test I added in this PR is based on @chhagedorn Test3.java. I was able to reproduce the issue on my end fairly easily.

Ah right, all good then. Please add @chhagedorn as co-contributor since he extracted that test.

> I wasn't able to reproduce the issue using Test.java, how often does it reproduce for you with the flags that you listed at the top of Test.java ?

Right, I just tried and it does not seem to reproduce anymore. Too bad but let's leave it out then.

File: src/hotspot/share/opto/output.cpp (hunk 2 above)
Comment (JohnTortugo) [reply src/hotspot/share/opto/output.cpp]: Done. Thanks Vladimir.

File: src/hotspot/share/opto/output.cpp [hunk 3]
```diff
@@ -1158,11 +1172,17 @@ void PhaseOutput::Process_OopMap_Node(MachNode *mach, int current_offset) {
...
-                         contains_as_scalarized_obj(jvms, sfn, objs, ov);
-          ov->set_root(is_root);
+          if (ov->is_root()) {
+            // Already flagged as 'root' by something else. We shouldn't change it
+            // to non-root in a younger JVMS because it may need to be alive in
+            // a younger JVMS.
+          } else {
+            bool is_root = locarray->contains(ov) ||
+                          exparray->contains(ov) ||
+                          contains_as_owner(monarray, ov) ||
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/output.cpp:1182]: Indentation is slightly off here.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too.

**Review by TobiHartmann — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]

**Pull request closed without merging**